src/gluonts/dataset/loader.py

```diff
@@ -148,121 +160,64 @@ def __init__(
         shuffle_for_training: bool = True,
         num_batches_for_shuffling: int = 10,
     ) -> None:
-        super().__init__(dataset, transform, batch_size, ctx, dtype)
-        self.num_batches_per_epoch = num_batches_per_epoch
-        self.shuffle_for_training = shuffle_for_training
-        self._num_buffered_batches = (
-            num_batches_for_shuffling if shuffle_for_training else 1
+        assert dataset, "empty dataset"
+
+        super().__init__(
+            dataset=itertools.cycle(dataset),
+            transform=transform,
+            batch_size=batch_size,
+            ctx=ctx,
+            dtype=dtype,
+            is_train=True,
         )
-        self._cur_iter: Optional[Iterator] = None
-        self._buffer = BatchBuffer(self.batch_size, ctx, dtype)
-
-    def _emit_batches_while_buffer_larger_than(
-        self, thresh
-    ) -> Iterator[DataBatch]:
-        if self.shuffle_for_training:
-            self._buffer.shuffle()
-        while len(self._buffer) > thresh:
-            yield self._buffer.next_batch()
-
-    def _iterate_forever(
-        self, collection: Iterable[DataEntry]
-    ) -> Iterator[DataEntry]:
-        # iterate forever over the collection, the collection must be non empty
-        while True:
-            try:
-                first = next(iter(collection))
-            except StopIteration:
-                raise Exception("empty dataset")
-            else:
-                for x in itertools.chain([first], collection):
-                    yield x
+
+        self.num_batches_per_epoch = num_batches_per_epoch
+
+        if shuffle_for_training:
+            self.stream = Shuffler(num_batches_for_shuffling, self.stream)
 
     def __len__(self) -> int:
         return self.num_batches_per_epoch
 
     def __iter__(self) -> Iterator[DataBatch]:
-        batch_count = 0
-        if self._cur_iter is None:
-            self._cur_iter = self.transform(
-                self._iterate_forever(self.dataset), is_train=True
-            )
-        assert self._cur_iter is not None
-        while True:
-            data_entry = next(self._cur_iter)
-            self._buffer.add(data_entry)
-            if (
-                len(self._buffer)
-                >= self._num_buffered_batches * self.batch_size
-            ):
-                for batch in self._emit_batches_while_buffer_larger_than(
-                    self.batch_size - 1
-                ):
-                    yield batch
-                    batch_count += 1
-                    if batch_count >= self.num_batches_per_epoch:
-                        return
+        return itertools.islice(self.batches, self.num_batches_per_epoch)
 
 
 class ValidationDataLoader(DataLoader):
-    """
```

Comment:
Why not keeping the docstring? We can improve it, in case we believe it’s not good

Reply:
I want to get rid of the class in the future. The only difference is whether `is_train` is set, and I think that should be handled outside.